src/oc/auth/async/slack_router.clj

```diff
@@ -0,0 +1,124 @@
+(ns oc.auth.async.slack-router
+  "Consume slack messages from SQS. This SQS queue is subscribed to the Slack
+   message SNS topic.
+  "
+  (:require
+   [clojure.core.async :as async :refer (<!! >!!)]
+   [cheshire.core :as json]
+   [oc.lib.sqs :as sqs]
+   [oc.lib.db.pool :as pool]
+   [oc.auth.config :as config]
+   [oc.auth.resources.team :as team-res]
+   [oc.auth.resources.slack-org :as slack-res]
+   [taoensso.timbre :as timbre]))
+
+;; ----- core.async -----
+
+(defonce slack-router-chan (async/chan 10000)) ; buffered channel
+
+(defonce slack-router-go (atom nil))
+
+(defn slack-event
+  "
+  Handle a message event from Slack. Ignore events that aren't threaded, or that are from us.
+  Idea here is to do very minimal processing and get a 200 back to Slack as fast as possible as this is a 'fire hose'
+  of requests. So minimal logging and minimal handling of the request.
+  Message events look like:
+  {'token' 'IxT9ZaxvjqRdKxYtWdTw21Xv',
+   'team_id' 'T06SBMH60',
+   'api_app_id' 'A0CHN2UDB',
+   'event' {'type' 'message',
+            'user' 'U06SBTXJR',
+            'text' 'Call me back here',
+            'thread_ts' '1494262410.072574',
+            'parent_user_id' 'U06SBTXJR',
+            'ts' '1494281750.011785',
+            'channel' 'C10A1P4H2',
+            'event_ts' '1494281750.011785'},
+    'type' 'event_callback',
+    'authed_users' ['U06SBTXJR'],
+    'event_id' 'Ev5B8YSYQ6',
+    'event_time' 1494281750}
+  "
+  [db-pool body]
+  (let [type (:type body)
+        token (:token body)
+        event (:event body)
+        channel (:channel event)
+        thread (:thread_ts event)]
+    ;; Token check
+    (if-not (= token config/slack-verification-token)
+
+      ;; Eghads! It might be a Slack impersonator!
+      (timbre/warn "Slack verification token mismatch, request provided:" token)
+
+      ;; Token check is A-OK
+      (when (= type "event_callback")
+        (let [event-type (:type event)]
+          (when (= event-type "tokens_revoked")
+            (let [slack-team-id (:team_id body)]
+              (pool/with-pool [conn db-pool]
+                (let [team (:team-id (first
+                                      (team-res/list-teams-by-index
+                                         conn
+                                         :slack-orgs
+                                         slack-team-id)))]
```

Comment:
@thepug i think you need a for here. You need to get all the carrot teams that have that slack org associated, they may be more then one if i got it right. For each of them remove the associated slack org and then remove the slack org itself from our db.
WYT?

Reply:
Yeah, I added a `doseq`.